Add mssql translations for two clock functions

ablack3 opened the pull request:
@hadley, @mgirlich. Here is what I'm thinking for two of the clock functions. Once I learn how to make well formed PRs I can add more but I wanted to see what you think about this one first. Can you review and discuss or make suggestions? Thanks!

hadley commented:
Yes, this what I was thinking 😄 

Tagging @DavisVaughan in case he has thoughts too,

DavisVaughan commented:
Seems useful to me!

mgirlich commented on R/backend-mssql.R:
```diff
@@ -350,6 +350,16 @@ simulate_mssql <- function(version = "15.0") {
+
+      # clock ---------------------------------------------------------------
+      add_days = function(x, n, ...) {
+        rlang::check_dots_empty(...)
```
You can omit the rlang namespace. The rlang package is so commonly used that it is usually imported.
And you either do `check_dots_empty()` or `check_dots_empty0(...)` but you must not pass `...` to `check_dots_empty()`.

```suggestion
        check_dots_empty()
```

mgirlich commented:
@ablack3 Thanks for you PR. Please add a NEWS entry mentioning the issue #1080 and yourself. And please modify your first message and mention that this closes #1080. This way the issue will be automatically closed when this PR is merged.

mgirlich commented:
@ablack3 Do you want to finish the PR? 😄 I'm happy to merge afterwards.

ablack3 commented:
> @ablack3 Do you want to finish the PR? 😄 I'm happy to merge afterwards.

Yes, I'll finish it this week.

hadley commented:
Yeah, I think `difftime()` is a reasonable translation.

hadley commented:
@ablack3 as you're working on this, it would be super useful if you'd take some notes so we could document the process of adding date-time wrappers for other backends (and then we could provide more expectations about how to express date time manipulations in R code for the user)

ablack3 commented:
> @ablack3 as you're working on this, it would be super useful if you'd take some notes so we could document the process of adding date-time wrappers for other backends (and then we could provide more expectations about how to express date time manipulations in R code for the user)

Yes of course. I'll write some documentation and share it for feedback. It would be nice to have one set of date manipulation functions that translate to many backends.

ablack3 commented:
Ok I think this PR is ready. I would expect that you'll have some recommendations so please let me know if you suggested edits to the PR. I will also add similar translations in the bigrquery package and duckdb package in separate PRs. Just wondering why the backends for bigquery and duckdb are not in the dbplyr package.

ablack3 commented on R/backend-postgres.R:
```diff
@@ -235,6 +235,29 @@ sql_translation.PqConnection <- function(con) {
+      },
+      add_years = function(x, n, ...) {
+        check_dots_empty()
+        sql_expr((!!x + !!n%*INTERVAL%'1 year'))
```
@mgirlich this works but is very awkward. Is there a better way to get the correct SQL here?

This is what I want to write but it is not a valid R expression
`sql_expr((!!x + !!n * INTERVAL'1 year'))`

The SQL should look like `\`column_name\` + 2 * INTERVAL'1 year'`

ablack3 replied:
@hadley is there a better way to write this expression? `!!x + !!n%*INTERVAL%'1 year'` seems strange but it's the only way I could come up with that worked.

ablack3 replied:
Ok I tried glue_sql2 instead. Hopefully I used it correctly.

ablack3 commented on R/backend-postgres.R:
```diff
@@ -235,6 +235,29 @@ sql_translation.PqConnection <- function(con) {
+        }
+
+        if (units[1] != "days") {
+          cli::cli_abort('The only supported value for {.arg units} on SQL backends is "days"')
```
I thought we could just support the 'days' unit to start. Keep it simple.

hadley reviewed (approved):
This seems like a good place to start. Do you want to add a news bullet and we can merge?

ablack3 commented:
> This seems like a good place to start. Do you want to add a news bullet and we can merge?

Ok I added a few more translations and added a news bullet. Thank you!
